### PR TITLE
runtime: Use a limited multi-threaded Tokio runtime in SGX

### PR DESCRIPTION
**Pull request by kostko**

### CODE REVIEW

**Review by peternose — COMMENTED** [review]
NIce, works as expected :+1: 

Tested on an SGX instance and can confirm, that the ephemeral secret replication deadlock problem, which occured if all key managers were restarted at the same time, is now gone.

File: runtime/src/enclave_rpc/client.rs [hunk 1]
```diff
@@ -204,301 +444,90 @@ impl RpcClient {
...
-                                .send((ctx, request, kind, rsp_tx, retries + 1))
-                                .await;
-                        }
-                    }
-                }
-
-                // Close stream after the client is dropped.
-                let _ = Self::close(inner).await;
-            });
+            tokio::spawn(controller.run());
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/client.rs:470]: The idea is that locking a mutex is more heavy than checking an atomic. And most of the time the mutex lock will not be necessary as we only need to initialize once.

[the remark this answers is not in the document]
Comment (peternose) [reply runtime/src/enclave_rpc/client.rs:470]: Ok, ignore the comment then.

File: runtime/src/enclave_rpc/demux.rs [hunk 2]
```diff
@@ -19,177 +23,243 @@ const DEFAULT_STALE_SESSION_TIMEOUT_SECS: u64 = 60;
...
+    /// Process a frame, returning the locked session guard and decoded message.
+    ///
+    /// Any data that needs to be transmitted back to the peer is written to the passed writer.
+    pub async fn process_frame<W: Write>(
+        &self,
+        data: Vec<u8>,
+        writer: W,
+    ) -> Result<
+        (
+            tokio::sync::OwnedMutexGuard<MultiplexedSession>,
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/demux.rs:212]: You would still need to keep the locked session as you need to also process the response, but actual RPC dispatch can only happen in the dispatcher.

File: runtime/src/enclave_rpc/client.rs [hunk 3]
```diff
@@ -204,301 +444,90 @@ impl RpcClient {
...
+    ///
+    /// This function panics if called within an asynchronous execution context.
     pub fn update_enclaves(&self, enclaves: Option<HashSet<EnclaveIdentity>>) {
-        let mut session = self.inner.session.lock().unwrap();
-        if session.builder.get_remote_enclaves() == &enclaves {
-            return;
-        }
-        session.builder = mem::take(&mut session.builder).remote_enclaves(enclaves);
-        session.reset();
+        let _ = self.cmdq.blocking_send(Command::UpdateEnclaves(enclaves));
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/client.rs:491]: This error probably cannot happen unless the controller panics as the controller will only get dropped once the client is dropped, but yeah will change to `unwrap`.

File: go/oasis-test-runner/scenario/e2e/runtime/runtime.go [hunk 4]
```diff
@@ -114,6 +116,7 @@ func newRuntimeImpl(name string, testClient TestClient) *runtimeImpl {
 	sc.Flags.String(cfgRuntimeBinaryDirDefault, "", "(no-TEE) path to the runtime binaries directory")
 	sc.Flags.String(cfgRuntimeBinaryDirIntelSGX, "", "(Intel SGX) path to the runtime binaries directory")
 	sc.Flags.String(cfgRuntimeLoader, "oasis-core-runtime-loader", "path to the runtime loader")
+	sc.Flags.String(cfgRuntimeProvisioner, "sandboxed", "the runtime provisioner: mock, unconfined, or sandboxed")
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/oasis-test-runner/scenario/e2e/runtime/runtime.go:119]: Yes? The above is only used for generating a fixture but one can always create a custom fixture and this field could be empty then.

[the remark this answers is not in the document]
Comment (kostko) [reply go/oasis-test-runner/scenario/e2e/runtime/runtime.go:119]: Oh I see, I thought you were referring to the first part. Did not know this `OASIS_UNSAFE_TESTS_NO_SANDBOX` existed. It seems strange to have it as an env variable, is this used anywhere?

[the remark this answers is not in the document]
Comment (kostko) [reply go/oasis-test-runner/scenario/e2e/runtime/runtime.go:119]: I'll remove it.

File: runtime/src/dispatcher.rs [hunk 5]
```diff
@@ -170,10 +170,13 @@ pub struct Dispatcher {
 impl Dispatcher {
     #[cfg(target_env = "sgx")]
     fn new_tokio_runtime() -> tokio::runtime::Runtime {
```
Comment (peternose) [inline runtime/src/dispatcher.rs:172]: In the future we could move this to `future.rs`, call it from `start_runtime` and pass it to dispatcher and protocol constructors.

**Review by peternose — APPROVED** [review]

File: runtime/src/dispatcher.rs (hunk 5 above)
Comment (kostko) [reply runtime/src/dispatcher.rs:172]: Oh, that's a good idea, yeah this part needs some refactoring.

**Pull request merged**